trie/nodeset.go

```diff
@@ -56,104 +54,110 @@ func (n *memoryNode) rlp() []byte {
+	owner      common.Hash            // the identifier of the trie
+	nodes      map[string]*memoryNode // the set of dirty nodes(inserted, updated, deleted)
+	leaves     []*leaf                // the list of dirty leaves
+	accessList map[string][]byte      // The list of accessed nodes, which records the original node value
 }
 
 // NewNodeSet initializes an empty node set to be used for tracking dirty nodes
-// from a specific account or storage trie. The owner is zero for the account
-// trie and the owning account address hash for storage tries.
-func NewNodeSet(owner common.Hash) *NodeSet {
+// for a specific account or storage trie. The owner is zero for the account trie
+// and the owning account address hash for storage tries. The provided accessList
+// represents the original value of accessed nodes, it can be optional but would
+// be beneficial for speeding up the construction of trie history.
+func NewNodeSet(owner common.Hash, accessList map[string][]byte) *NodeSet {
+	// Don't panic for lazy users.
+	if accessList == nil {
+		accessList = make(map[string][]byte)
+	}
 	return &NodeSet{
-		owner: owner,
-		updates: &nodesWithOrder{
-			nodes: make(map[string]*nodeWithPrev),
-		},
-		deletes: make(map[string][]byte),
+		owner:      owner,
+		nodes:      make(map[string]*memoryNode),
+		accessList: accessList,
 	}
 }
 
-/*
-// NewNodeSetWithDeletion initializes the nodeset with provided deletion set.
-func NewNodeSetWithDeletion(owner common.Hash, paths [][]byte, prev [][]byte) *NodeSet {
-	set := NewNodeSet(owner)
-	for i, path := range paths {
-		set.markDeleted(path, prev[i])
+// forEachWithOrder iterates the dirty nodes with the specified order.
+// If topToBottom is true:
+//
+//	then the order of iteration is top to bottom, left to right.
+//
+// If topToBottom is false:
+//
+//	then the order of iteration is bottom to top, right to left.
+func (set *NodeSet) forEachWithOrder(topToBottom bool, callback func(path string, n *memoryNode)) {
+	var paths sort.StringSlice
+	for path := range set.nodes {
+		paths = append(paths, path)
+	}
+	if topToBottom {
+		paths.Sort()
+	} else {
+		sort.Sort(sort.Reverse(paths))
+	}
+	for _, path := range paths {
+		callback(path, set.nodes[path])
 	}
-	return set
 }
-*/
 
 // markUpdated marks the node as dirty(newly-inserted or updated) with provided
 // node path, node object along with its previous value.
-func (set *NodeSet) markUpdated(path []byte, node *memoryNode, prev []byte) {
-	set.updates.order = append(set.updates.order, string(path))
-	set.updates.nodes[string(path)] = &nodeWithPrev{
-		memoryNode: node,
-		prev:       prev,
-	}
+func (set *NodeSet) markUpdated(path []byte, node *memoryNode) {
+	set.nodes[string(path)] = node
 }
 
 // markDeleted marks the node as deleted with provided path and previous value.
-func (set *NodeSet) markDeleted(path []byte, prev []byte) {
-	set.deletes[string(path)] = prev
+func (set *NodeSet) markDeleted(path []byte) {
```

Comment:
This method is only used in tests. Nuke?

Reply:
Yep, originally I want to capture the deleted nodes from trie. But the relevant functionality is removed. I am fine with nuking it out.